### DISCUSSION TITLE
How can I get more details about the task call stack

### DISCUSSION THREAD

**Discussion by sukhmel** [tokio-rs/console#599] [Q&A]
Hi,

I have a bunch of async functions that I pass to a lib for execution, joining, and printing error result if any. When I connect to the application with `tokio-console` I only see it point to the location in library code that does this:

```rust
.map(|Handler { label, future }| (label, tokio::spawn(future)))
```

location reads: `<cargo>/library_name-0.1.0/src/lib.rs:99:54`

Is it possible to get more details beyond that? I'm going to try to spawn all tasks from my code and give them names, too, but I can't quite understand what is the limitation (or error with my usage).

Besides, I can't seem to get all the fancy graphs like on the screenshots, what I have is quite scarce, is that also some kind of error on my side?

```
Warnings
/!\ This task has woken itself for more than 50% of its total wakeups (96%) 

Task                                                                  Waker
ID: 12 IDLE                                                           Current wakers: 2 (clones: 798, drops: 796) 
Target: tokio::task                                                   Woken: 12517 times, last woken: 81.94688888s ago 
Location: <cargo>/library_name-0.1.0/src/lib.rs:99:54                 Self Wakes: 12133 times (96%)
Total Time: 43m51s
Busy: 15.50s (0.59%)
Scheduled: 117.82ms (0.00%)    
Idle: 43m36s (99.41%) 


Poll Times Percentiles 
p10: 1.06ms 
p25: 1.07ms
p50: 1.11ms
p75: 1.16ms  
p90: 1.23ms
p95: 1.46ms 
p99: 4.52ms 

Sched Times Percentiles  
p10: 7.52us  
p25: 8.00us  
p50: 8.70us  
p75: 9.41us 
p90: 10.30us  
p95: 11.58us 
p99: 35.58us  
 
Fields
kind=task
size.bytes=16
```

**Comment by hds** [comment]
The `console-subscriber` uses the [`panic::Location`](https://doc.rust-lang.org/std/panic/struct.Location.html) to get the location where a task was spawned from. That is wherever the `tokio::spawn` call was performed unless that function is annotated with [`#[track_caller]`](https://doc.rust-lang.org/reference/attributes/codegen.html#the-track_caller-attribute), in which case the panic location is where **that** function was called.

However, there's no way to annotate a closure, so in the case of the library you're using, it isn't possible to get the location in your own code. If that `.map` was instead in a `for` loop it would be possible.

(For some more details, I wrote a bit about the `#[track_caller]` attribute here: https://hegdenu.net/posts/track-caller/)

As for the pretty graphs, Tokio Console uses the language environment to decide whether it can show those graphs or not, it needs to be something UTF-8 compatible. You can overwrite it on the command line using `--lang=en_us.UTF-8` (or whatever language, so long as it has the `UTF-8` suffix).

(P.S. I hope you're fine after the dana. Hi from Granada 👋)

**Reply by sukhmel** [reply]
Thanks, I will try with annotation and maybe push some changes to the library.

I'm good, thank you. For the most part, flooding happened outside of the city, some of my friends and colleagues were affected, so we're helping out a bit

**Reply by sukhmel** [reply]
I returned to this problem again and looked into how `#[track_caller]` could be used, and I'm not entirely sure if that's even possible to do. 

To elaborate, I have a number of `BoxFuture`s that contain handlers and are spawned with `tokio::task::Builder::new().name(&label).spawn(future)`. This way I can see which handler is running and how much, but it doesn't provide any details about where exactly execution is. As far as I understand, track caller can't be used with any async functions, so even if I replace anonymous functions making `BoxFuture` with normal functions, those will be async, and will not make much difference.

I'm starting to think that it's not possible to see that information really, maybe I should look into some other tools for that? My idea was to use `tokio-console` to look into state of an application when it hangs somewhat, but it feels like the only thing I will see is that tasks are not doing much and not where
